### PR TITLE
Misc optimizations for Blob / BlobDispatcher and URLBuilder

### PR DESCRIPTION
**Pull request by andyHa**
Please revise the individual commints for elaborate change descriptions.

This main tipping point for this refactoring was: [SIRI-725](https://scireum.myjetbrains.com/youtrack/issue/SIRI-725)

Breaking changes:

Some methods and return types in URLBuilder have been removed/renamed:
- eternallyValid() is replaced by reusable()
- delayedResolved() is replaced by delayedResolve()
- The url in URLBuilder.UrlResult is now non-optional as its value is dictated by the UrlType already

For Blob:
- deliver() was removed as all traffic is sent through the BlobDispatcher

### CODE REVIEW

**Review by jakobvogel — APPROVED** [review]

File: src/main/java/sirius/biz/storage/layer2/BlobDispatcher.java [hunk 1]
```diff
@@ -384,19 +388,37 @@ private void virtualDelivery(WebContext request,
...
         Response response = request.respondWith();
+
         if (cacheable) {
-            response.cached();
+            // If a virtual request is marked as cacheable, we try to redirect to the proper physical blob key
+            // as this will remain in cache much longer (and the redirect itself will also be cached). The additional
+            // HTTP round-trip for the redirect shouldn't hurt too much, as it is most probably optimized away due to
+            // keep-alive. However, using a physical delivery with infinite cache settings will enable any downstream
+            // reverse-proxies to maximize their cache utilization...
+            if (attemptRedirectToPhysical(response,
```
Comment (idlira) [inline src/main/java/sirius/biz/storage/layer2/BlobDispatcher.java:403]: Just a design topic: I don't like the approach of changing the response inside the method (makes reading it more complex). Would make the method return the url, and if non empty set the response/redirect hier.

File: src/main/java/sirius/biz/storage/layer2/Blob.java [hunk 2]
```diff
@@ -134,7 +134,10 @@ public interface Blob {
      * Delivers the data of this blob into the given HTTP response.
      *
      * @param response the HTTP response to populate
+     * @deprecated use {@link URLBuilder} and let the {@link BlobDispatcher} perform the heavy lifting. If a response
+     * is to be fulfilled, redirect to the URL generated by the url builder...
      */
+    @Deprecated
```
Comment (idlira) [inline src/main/java/sirius/biz/storage/layer2/Blob.java:140]: Kill right now?

File: src/main/java/sirius/biz/storage/layer2/URLBuilder.java [hunk 3]
```diff
@@ -243,21 +242,25 @@ public URLBuilder withBaseURL(String baseURL) {
      * Note that this implies that a {@link #reusable() reusable} URL is generated.
      *
      * @return the builder itself for fluent method calls
+     * @deprecated can be replaced by {@link #reusable()} - which should have been invoked in the first place and which
+     * now automatically generates eternally valid URLs.
      */
+    @Deprecated
```
Comment (idlira) [inline src/main/java/sirius/biz/storage/layer2/URLBuilder.java:276]: Kill right now?

File: src/main/java/sirius/biz/storage/layer2/URLBuilder.java [hunk 4]
```diff
@@ -269,6 +269,14 @@ public URLBuilder reusable() {
         return this;
     }
 
+    /**
+     * @deprecated use {@link #delayResolve()} which is the same method except for the typo in the name
+     */
+    @Deprecated
```
Comment (idlira) [inline src/main/java/sirius/biz/storage/layer2/URLBuilder.java:303]: Kill right now?

**Review by idlira — CHANGES_REQUESTED** [review]

File: src/main/java/sirius/biz/storage/layer2/BlobDispatcher.java [hunk 5]
```diff
@@ -411,4 +429,31 @@ private void virtualDelivery(WebContext request,
                              response,
                              request::markAsLongCall);
     }
+
+    private boolean attemptRedirectToPhysical(@Nullable Response response,
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer2/BlobDispatcher.java:433]: The Response parameter is not actually nullable in the Code below

File: src/main/java/sirius/biz/storage/layer1/ObjectStorageSpace.java [hunk 6]
```diff
@@ -481,6 +482,28 @@ public void deliver(Response response, String objectId, boolean largeFileExpecte
...
+    /**
+     * Determines if a special delivery for very large files should be attempted.
+     * <p>
+     * For very large files (multiple GBs), we use special URLs generated by the
+     * {@link sirius.biz.storage.layer2.URLBuilder} and handled by the {@link sirius.biz.storage.layer2.BlobDispatcher}.
+     * These contain the "/xxl/" marker so that downstream caches like <tt>Varnish</tt> don't even attempt to put the
+     * data in their cache. These files therefore are piped through which may lead to increasingly large internal
+     * buffers as an upstream storage like <b>S3</b> might send data faster than we pump through to slower clients
+     * like mobile users. We therefore use a blocking/stream based approach in a thread pool of limited size. However,
+     * in the presence of a <tt>Range</tt> header, the date to transfer is most probably way smaller and also the stream
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/ObjectStorageSpace.java:494]:
```suggestion
     * in the presence of a <tt>Range</tt> header, the data to transfer is most probably way smaller and also the stream
```

File: src/main/java/sirius/biz/storage/layer1/ObjectStorageSpace.java [hunk 7]
```diff
@@ -481,6 +482,28 @@ public void deliver(Response response, String objectId, boolean largeFileExpecte
...
+     * implementation cannot satisfy such requests. We therefore only use this approach is a large file is expected and
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/ObjectStorageSpace.java:495]:
```suggestion
     * implementation cannot satisfy such requests. We therefore only use this approach if a large file is expected and
```

File: src/main/java/sirius/biz/storage/layer1/ObjectStorageSpace.java [hunk 8]
```diff
@@ -481,6 +482,28 @@ public void deliver(Response response, String objectId, boolean largeFileExpecte
...
+     * no <tt>Range</tt> header is present.
+     *
+     * @param response          the response to handle
+     * @param largeFileExpected a flag which determines if a large file is expected at all
+     * @return <tt>true</tt> if the delivery should be handled in a separate thread pool and a non-blocking fashion,
+     * <tt>false</tt> otherwise
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/ObjectStorageSpace.java:501]:
```suggestion
     * @return <tt>true</tt> if the delivery should be handled in a separate thread pool and a non-blocking fashion
```
In my opinion this its redundant to tell that a boolean parameter can also be false :)

File: src/main/java/sirius/biz/storage/layer2/BlobDispatcher.java (hunk 1 above)
Comment (sabieber) [reply src/main/java/sirius/biz/storage/layer2/BlobDispatcher.java:403]: In general I don't like the pattern of altering a parameter inside a method, too. But in the case of the response here it is not really "altering" the response but completing it?
But I also agree that keeping all actually response calls in this method and not extract them away would simplify understanding what can be the response we send out.

File: src/main/java/sirius/biz/storage/layer2/URLBuilder.java [hunk 9]
```diff
@@ -76,10 +76,10 @@ public class URLBuilder {
      * Contains the {@linkplain #buildUrlResult() generated URL} and an {@linkplain UrlType indicator} if it is a
      * virtual, physical, the fallback URL, or empty.
      *
-     * @param url     the URL generated by this url builder wrapped as an Optional
+     * @param url     the URL generated by this url builder
      * @param urlType the type of URL that was generated
      */
-    public record UrlResult(Optional<String> url, UrlType urlType) {
+    public record UrlResult(String url, UrlType urlType) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer2/URLBuilder.java:82]: This is a breaking change? Please list it in the PR description

**Review by sabieber — APPROVED** [review]

**Review by jakobvogel — APPROVED** [review]

**Review by idlira — APPROVED** [review]

**Review by mko-sci — APPROVED** [review]

**Pull request merged**